File: Practica 6/Vector3D.java
class Vector3D {
    private double x, y, z;

    public Vector3D(double x, double y, double z) {
        this.x = x;
        this.y = y;
        this.z = z;
    }

    public Vector3D sumar(Vector3D otro) {
        return new Vector3D(this.x + otro.x, this.y + otro.y, this.z + otro.z);
    }

    public Vector3D multiplicar(double escalar) {
        return new Vector3D(this.x * escalar, this.y * escalar, this.z * escalar);
    }

    public double norma() {
        return Math.sqrt(x * x + y * y + z * z);
    }

    public Vector3D normalizar() {
        double norma = this.norma();
        return (norma == 0) ? new Vector3D(0, 0, 0) : new Vector3D(x / norma, y / norma, z / norma);
    }

    public double productoEscalar(Vector3D otro) {
        return this.x * otro.x + this.y * otro.y + this.z * otro.z;
    }

    public Vector3D productoVectorial(Vector3D otro) {
        return new Vector3D(
            this.y * otro.z - this.z * otro.y,
            this.z * otro.x - this.x * otro.z,
            this.x * otro.y - this.y * otro.x
        );
    }

    public String toString() {
        return "(" + x + ", " + y + ", " + z + ")";
    }

    public static void main(String[] args) {
        Vector3D v1 = new Vector3D(1, 2, 3);
        Vector3D v2 = new Vector3D(2, 4, 6);

        System.out.println("Suma de vectores: " + v1.sumar(v2));
        System.out.println("Multiplicación por escalar: " + v1.multiplicar(3));
        System.out.println("Norma del vector: " + v1.norma());
        System.out.println("Vector normalizado: " + v1.normalizar());
        System.out.println("Producto escalar: " + v1.productoEscalar(v2));
        System.out.println("Producto vectorial: " + v1.productoVectorial(v2));
    }
}

File: Practica 6/vector3d.py
import numpy as np

class Vector3D:
    def __init__(self, x, y, z):
        self.x = x
        self.y = y
        self.z = z
    
    def __add__(self, otro):
        return Vector3D(self.x + otro.x, self.y + otro.y, self.z + otro.z)
    
    def __mul__(self, escalar):
        return Vector3D(self.x * escalar, self.y * escalar, self.z * escalar)
    
    def norma(self):
        return np.sqrt(self.x**2 + self.y**2 + self.z**2)
    
    def normalizar(self):
        norma = self.norma()
        if norma == 0:
            return Vector3D(0, 0, 0)
        return Vector3D(self.x / norma, self.y / norma, self.z / norma)
    
    def producto_escalar(self, otro):
        return self.x * otro.x + self.y * otro.y + self.z * otro.z
    
    def producto_vectorial(self, otro):
        return Vector3D(
            self.y * otro.z - self.z * otro.y,
            self.z * otro.x - self.x * otro.z,
            self.x * otro.y - self.y * otro.x
        )
    
    def __repr__(self):
        return f"Vector3D({self.x}, {self.y}, {self.z})"


v1 = Vector3D(1, 2, 3)
v2 = Vector3D(2, 4, 6)
print("Suma de vectores:", v1 + v2)
print("Multiplicación por escalar:", v1 * 3)
print("Norma del vector:", v1.norma())
print("Vector normalizado:", v1.normalizar())
print("Producto escalar:", v1.producto_escalar(v2))
print("Producto vectorial:", v1.producto_vectorial(v2))
